doc/source/templates/04_finetuning_llms_with_deepspeed/finetune_hf_llm.py

```diff
@@ -0,0 +1,569 @@
+        perplexity = math.exp(eval_loss)
+    except OverflowError:
+        perplexity = float("inf")
+    return perplexity, eval_loss
+
+
+def checkpoint_model(checkpoint_folder, ckpt_id, model, epoch, last_global_step, **kwargs):
+    """Utility function for checkpointing model + optimizer dictionaries
+    The main purpose for this is to be able to resume training from that instant again.
+    """
+    checkpoint_state_dict = {
+        "epoch": epoch,
+        "last_global_step": last_global_step,
+    }
+    # Add extra kwargs too
+    checkpoint_state_dict.update(kwargs)
+    
+    # In here model will be a DeepspeedEngine object
+    model.save_checkpoint(checkpoint_folder, ckpt_id, checkpoint_state_dict)
+    status_msg = f"checkpointing: checkpoint_folder={checkpoint_folder}, ckpt_id={ckpt_id}"
+    print(status_msg)
+
+def training_function(kwargs: dict):
+    os.environ["OMP_NUM_THREADS"] = str(
+        session.get_trial_resources().bundles[-1].get("CPU", 1)
+    )
+
+    print("training_function called")
+
+    config = kwargs["config"]
+    args = argparse.Namespace(**kwargs["args"])
+    special_tokens = kwargs.get("special_tokens", [])
+    model_id = config["model_name"]
+
+    # Sample hyper-parameters for learning rate, batch size, seed and a few other HPs
+    lr = config["lr"]
+    num_epochs = int(config["num_epochs"])
+    seed = int(config["seed"])
+    batch_size = int(config["batch_size"])
+    gradient_accumulation_steps = int(config["gradient_accumulation_steps"])
+
+
+    # Get deepspeed config to setup the batch size per device
+    ds_plugin = config["ds_plugin"]
+    ds_plugin.hf_ds_config.config["train_micro_batch_size_per_gpu"] = batch_size
+
+    # Initialize accelerator
+    accelerator = Accelerator(
+        deepspeed_plugin=ds_plugin,
+        gradient_accumulation_steps=gradient_accumulation_steps,
+        mixed_precision=args.mx,
+    )
+    
+    set_seed(seed)
+    
+    # train_ds is the local shard for this model
+    train_ds = session.get_dataset_shard("train")
+    valid_ds = session.get_dataset_shard("valid")
+    
+    train_ds_len = len(list(train_ds.iter_batches(batch_size=1)))
+    valid_ds_len = len(list(valid_ds.iter_batches(batch_size=1)))
+    
+    tokenizer = get_tokenizer(model_name=args.model_name, special_tokens=special_tokens)
+    collate_partial = functools.partial(
+        collate_fn, 
+        tokenizer=tokenizer,
+        block_size=config["block_size"], 
+        device=accelerator.device
+    )
+    
+    # Get the trial directory from Ray Train
+    # This will be local to every node (and will get synced to remote storage if provided.)
+    ckpt_path = session.get_trial_dir()
+
+    pretrained_path = get_pretrained_path(model_id)    
+    print(f"Loading model from {pretrained_path} ...")
+    s = time.time()
+    model = AutoModelForCausalLM.from_pretrained(
+        pretrained_path,
+        trust_remote_code=True, 
+        torch_dtype=torch.bfloat16,
```

Comment:
Should this be based on `args.mx`?

Reply:
Actually no. You load the model in bf16 so that the consumed RAM is at least in 16 bits. Then if the model needs to be trained in full precision you shared it in fp32